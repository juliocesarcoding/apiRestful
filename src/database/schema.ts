import db from "./db";

// Create a table with the specified schema
export const createTable = `
CREATE TABLE IF NOT EXISTS movies (
id INTEGER PRIMARY KEY AUTOINCREMENT,
    year INTEGER,
    title TEXT,
    studios TEXT,
    producers TEXT,
    winner BOOLEAN
);
`;

// Execute the SQL command
db.exec(createTable);

// Close the database connection
db.close();
