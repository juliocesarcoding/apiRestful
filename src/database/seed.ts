// import { initDatabase as db } from "./index";
// import fs from "fs";
// import path from "path";
// import csvParser from "csv-parser";
// import { Database } from "better-sqlite3";

// export const insertMoviesFromCsv = async (db: Database) => {
//  const csvFilePath = path.resolve(__dirname, "../../public/Movielist.csv");

//  return new Promise<void>((resolve, reject) => {
//      fs
//          .createReadStream(csvFilePath)
//          .pipe(csvParser({ separator: "," }))
//          .on("data", async (row) => {
//              const winner = row.winner === "yes" ? 1 : 0;
//              const insert = db.prepare(
//                  "INSERT INTO movies (year, title, studios, producers, winner) VALUES (?, ?, ?, ?, ?)"
//              );
//              db.transaction((rows) => {
//                  for (const row of rows)
//                      insert.run(row.year, row.title, row.studios, row.producers, row.winner);
//              })
//        .on("end", () => {
//          console.log("CSV file successfully processed");
//         resolve();
//        })
//        .on("error", (error) => {
//         console.error("Error processing CSV file", error);
//         reject(error);
//        });
//      });
//    }
type MovieRow = {
 year: number;
 title: string;
 studios: string;
 producers: string;
 winner: number;
}[];

import fs from "fs";
import path from "path";
import csvParser from "csv-parser";
import { Database } from "better-sqlite3";

export const insertMoviesFromCsv = async (db: Database) => {
 const csvFilePath = path.resolve(__dirname, "../../public/Movielist.csv");

 return new Promise<void>((resolve, reject) => {
  const insert = db.prepare(
   "INSERT INTO movies (year, title, studios, producers, winner) VALUES (?, ?, ?, ?, ?)"
  );
  const transaction = db.transaction((rows) => {
   for (const row of rows) {
    insert.run(row.year, row.title, row.studios, row.producers, row.winner);
   }
  });

  const rows: MovieRow = [];
  fs
   .createReadStream(csvFilePath)
   .pipe(csvParser({ separator: ";" }))
   .on("data", (row) => {
    const winner = row.winner === "yes" ? 1 : 0;
    rows.push({ ...row, winner });
   })
   .on("end", () => {
    transaction(rows);
    console.log("rows", rows);
    console.log("CSV file successfully processed");
    resolve();
   })
   .on("error", (error) => {
    console.error("Error processing CSV file", error);
    reject(error);
   });
 });
};
