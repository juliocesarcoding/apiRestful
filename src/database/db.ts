import Database from "better-sqlite3";
import { insertMoviesFromCsv } from "./seed";

const db = new Database(":memory:");

export const initDatabase = async () => {
 const createTable = `
CREATE TABLE IF NOT EXISTS movies (
id INTEGER PRIMARY KEY AUTOINCREMENT,
    year INTEGER,
    title TEXT,
    studios TEXT,
    producers TEXT,
    winner BOOLEAN
);
`;
 db.exec(createTable);
 await insertMoviesFromCsv(db);
};

export default db;
