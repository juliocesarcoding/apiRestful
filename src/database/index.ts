import Database from "better-sqlite3";
import { createTable } from "./schema";
import { insertMoviesFromCsv } from "./seed";
const db = new Database(":memory:");

export const initDatabase = async () => {
 db.exec(createTable);
 console.log("Database initialized");
};

export default db;
