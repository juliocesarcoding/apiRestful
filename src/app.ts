import express, { ErrorRequestHandler, Request, Response } from "express";
import cors from "cors";
import main from "./routes/index";

import { initDatabase } from "./database";
const server = express();
server.use(cors());
server.use(express.urlencoded({ extended: true }));
server.use("/api", main);
server.use((req: Request, res: Response) => {
 res.status(404);
 res.json({ error: "Endpoint not found." });
});

const errorHandler: ErrorRequestHandler = (err, req, res, next) => {
 res.status(400);
 console.log(err);
 res.json({ error: "An error occured. Bad Request" });
};
// initDatabase();
server.use(errorHandler);
server.use(express.json());

export default server;
