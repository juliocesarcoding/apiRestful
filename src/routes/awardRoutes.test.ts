import request from "supertest";
import app from "../app";
import { initDatabase } from "../database/db";

describe("Test award endpoint", () => {
 beforeAll(async () => {
  await initDatabase();
 });

 it("Should return 200 and an object of awards", async () => {
  const response = await request(app).get("/api/awards");
  expect(response.status).toBe(200);
  expect(response.body).toMatchSnapshot();
 });
 it("Should ");
});
