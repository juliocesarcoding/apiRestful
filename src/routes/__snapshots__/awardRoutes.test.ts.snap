// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test award endpoint Should return 200 and an object of awards 1`] = `
{
  "max": [
    {
      "followingWin": 2015,
      "interval": 13,
      "previousWin": 2002,
      "producer": "Matthew Vaughn",
    },
  ],
  "min": [
    {
      "followingWin": 1991,
      "interval": 1,
      "previousWin": 1990,
      "producer": "Joel Silver",
    },
  ],
}
`;
